### PR TITLE
New mismatch pair creation strategy

### LINKED ISSUE #1911: Minor KTO mis-estimation of KL term on odd batch sizes
In kto_trainer.py, trl calculates the KL term by reversing the labels. That'll usually work, except on odd batch sizes, in which case the middle will be a match.

```
def _get_kl_dataset(batch: Dict[str, List[Any]]) -> Dict[str, List[Any]]:
    """Creates mismatched pairs of prompts and completions for the KL dataset by reversing the order of completions."""
    batch["answer_input_ids"] = batch["answer_input_ids"][::-1]
    batch["answer_attention_mask"] = batch["answer_attention_mask"][::-1]
    return batch
```

The halo repo does it with an offset, which works with any >1 odd batch sizes.
```
                    # for estimating the KL term, match up x and y' that are not corresponding input-output pairs in the data
                    # for x_i, get a mismatched y' by just picking the subsequent y_{i+1} in the batch (desirable/undesirable status does not matter)
                    # the respective input IDs, attention mask, and so on will be prefixed by the term KL
                    indices = list(range(1, len(batch))) + [0]
                    for i in range(len(batch)):
                        batch[i].update(self.tokenize_batch_element(
                            batch[i]['prompt_text'],
                            batch[indices[i]]['target_text'],
                            batch[i]['truncation_mode'],
                            prefix='KL'
                        ))
```

`https://github.com/ContextualAI/HALOs/blob/main/dataloader.py`

Not a huge bug, but thought it'd be worth pointing out

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes #1911

With the current `_get_kl_dataset` in [`kto_trainer.py`](https://github.com/huggingface/trl/blob/main/trl/trainer/kto_trainer.py) you can have this situation when the batch size is odd:

```python
from typing import Any, Dict, List

def _get_kl_dataset(batch: Dict[str, List[Any]]) -> Dict[str, List[Any]]:
    """Creates mismatched pairs of prompts and completions for the KL dataset by reversing the order of completions."""
    batch["answer_input_ids"] = batch["answer_input_ids"][::-1]
    batch["answer_attention_mask"] = batch["answer_attention_mask"][::-1]
    return batch


input_ids = ["example_0", "example_1", "example_2"]
attention_mask = ["attention_mask_0", "attention_mask_1", "attention_mask_2"]

batch = {"answer_input_ids": input_ids, "answer_attention_mask": attention_mask}
kl_dataset = _get_kl_dataset(batch)["answer_input_ids"]

print(input_ids) # ['example_0', 'example_1', 'example_2']
print(kl_dataset) # ['example_2', 'example_1', 'example_0']
```

`example_1` matches, while it shouldn't.

To solve this, we add a +1 offset instead:

```python
from typing import Any, Dict, List

def _get_kl_dataset(batch: Dict[str, List[Any]]) -> Dict[str, List[Any]]:
    """Creates mismatched pairs of prompts and completions for the KL dataset by adding a +1 offset to the order of completions."""
    batch["answer_input_ids"]  = [batch["answer_input_ids"][-1]] + batch["answer_input_ids"][:-1]
    batch["answer_attention_mask"] = [batch["answer_attention_mask"][-1]] + batch["answer_attention_mask"][:-1]
    return batch


input_ids = ["example_0", "example_1", "example_2"]
attention_mask = ["attention_mask_0", "attention_mask_1", "attention_mask_2"]

batch = {"answer_input_ids": input_ids, "answer_attention_mask": attention_mask}
kl_dataset = _get_kl_dataset(batch)["answer_input_ids"]

print(input_ids) # ['example_0', 'example_1', 'example_2']
print(kl_dataset) # ['example_2', 'example_0', 'example_1']
```

Now, all the examples mismatch.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?: locally

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by kashif — APPROVED** [review]
cool!
